src/NServiceBus.Core/Routing/DistributionStrategy.cs

```diff
@@ -24,6 +27,14 @@ protected DistributionStrategy(string endpoint, DistributionStrategyScope scope)
         public abstract string SelectReceiver(string[] receiverAddresses);
 
         /// <summary>
+        /// Selects a destination instance for a message from all known addresses of a logical endpoint.
+        /// </summary>
+        public virtual Task<string> SelectReceiver(string[] receiverAddresses, IOutgoingContext outgoingContext)
```

Comment:
I'm just thinking whether we really need this API to support async implementations. In the routing logic we tried to avoid async whenever possible for performance reasons. The mindset was that async operations should happen on the side.

e.g. we do that with the instance mapping file or file based routing. Files are reloaded in a regular interval, but reading files on every outgoing message in definitely not something which would yield good performance. Not sure about your use case, but I have a feeling this applies here as well.